W-10900429: Update mule-high-availability-ha-clusters.adoc (#2181)
mulesoft/docs-mule-runtime-3: modules/ROOT/pages/mule-high-availability-ha-clusters.adoc

Mule Enterprise Edition supports scalable clustering to provide high availability (HA) for applications.

A cluster is a set of Mule runtime engines that acts as a unit. In other words, a cluster is a virtual server composed of multiple nodes (Mule runtime engines). The nodes in a cluster communicate and share Object Store and VM queue data through a distributed shared memory grid. This means that the data is replicated across memory in different machines. 

This shared memory grid, replicated state applies to: 

> * persistent and non-persistent Object stores 

Reviewer: Object stores -> Object Store instances ?

Author: Don't think we really use the term "instances" elsewhere - the Object Store connector for example just says "Object store" for stuff like "default object store". So instances I think is redundant.

> * persistent and transient VM queues

Reviewer 2: ```suggestion
* persistent and transient VM queues
* Mule runtime LockFactory
```

> 

Reviewer 3: Cluster also overrides the Mule Runtime LockFactory, so if we will enumerate the things that are shared, then we should mention the locks too.

== The Benefits of Clustering

> By default, clustering Mule runtime engines ensures high system availability. If a Mule runtime engine node becomes unavailable due to failure or planned downtime, another node in the cluster can assume the workload and continue to process messages from VM queues as well as continue to service other requests. The following figure illustrates the processing of incoming messages by a cluster of two nodes. Notice that the processing load is balanced across nodes: Node 1 processes message 1 while Node 2 simultaneously processes message 2.

Reviewer: ```suggestion
By default, clustering Mule runtime engines ensures high system availability. If a Mule runtime engine node becomes unavailable due to failure or planned downtime, another node in the cluster can assume the workload and continue to process messages from VM queues and to service other requests. The following figure illustrates the processing of incoming messages by a cluster of two nodes. Notice that the processing load is balanced across nodes: Node 1 processes message 1 while Node 2 simultaneously processes message 2.
```

Author: Looks good!

image::failovernofail.png[FailoverNoFail]

> If one node fails, the other available nodes pick up the work of the failing node. This would be via an external load balancer redirecting the share of traffic to the still active node, or via use of VM queues to enable the active nodes to continue to process in-flight messages. As shown in the following figure, if Node 2 fails, Node 1 processes both message 1 and message 2.

Reviewer: ```suggestion
If one node fails, the other available nodes pick up the work of the failing node. An external load balancer redirects the failing node's share of traffic to an active node, or through VM queues to enable the active nodes to continue to process in-flight messages. As shown in the following figure, if Node 2 fails, Node 1 processes both message 1 and message 2.
```

Author: sounds good

image::failovernode2fail.png[FailoverNode2Fail]

> Because all nodes in a cluster of Mule runtime engines are "active active" and can process messages simultaneously, clusters can also improve performance and scalability. Compared to a single node instance, clusters can support more users or improve application performance by sharing the workload across multiple nodes or by adding nodes to the cluster. Note that not all applications will perform better with horizontally scaling or clustering - as it depends on having work of a nature that can be shared with those additional nodes, and some applications with heavy dependency on Object Stores might suffer from the work to replicate/coordinate access to the data within the shared memory grid. 

Reviewer: ```suggestion
Because all nodes in a cluster of Mule runtime engines are "active active" and can process messages simultaneously, clusters can also improve performance and scalability. Compared to a single node instance, clusters can support more users or improve application performance by sharing the workload across multiple nodes or by adding nodes to the cluster. Note that not all applications perform better when horizontally scaling or clustering. Performance depends on the nature of the work to be shared with additional nodes. Performance of some applications with a heavy dependency on Object Stores can degrade because of the work required to replicate or coordinate access to the data within the shared memory grid. 
```

Author: The term "Ease of sharing work" kinda feels a bit more human-ish than the wording for an app server. We're talking stuff that either _is or is not_  shared (the old Yoda "do or do not, there is no try"). There's no relative difficulty involved - it's a binary yes or no - it's local or it's shared. 

And the term instances - I reckon going simply with "an object store" or "object stores" rather than object store instances. But not really that big a deal either way I don't think it's outrageous to say instances.. Although there may be an argument to talk about "partition" perhaps, although would need engineering to clarify "object store" vs "object store instance" vs "object store partition".

Reviewer 3: Agree with the "ease of sharing work" comment from Nathan. The "nature" phrasing sounds more appropriate to me

Reviewer 2: @eze210 
“Performance depends on the nature of the work to be shared with additional nodes.”

All Mule instances get the same TCP requests and then process duplicate messages.

* JMS topics. +

> All Mule instances connect to the same JMS topic, which might lead to the repeated processing of messages when scaling the non-clustered Mule instance horizontally. To the JMS broker the instances appear as separate subscribers and will all get a copy of each message, which is rarely the required behavior. Although there is a "shared subscriber" configuration available which might also address this requirement to instruct the JMS broker to treat all instances as a combined subscriber and to give them each separate messages.

Reviewer: ```suggestion
All Mule instances connect to the same JMS topic, which might lead to the repeated processing of messages when scaling the non-clustered Mule instance horizontally. To the JMS broker, the instances appear as separate subscribers, all of which get a copy of each message. This behavior is rarely required, but a "shared subscriber" configuration is available to instruct the JMS broker to treat all instances as a combined subscriber and to give them each separate messages.
```

Author: The proposed split of the sentences with "but" has changed the meaning: "This behaviour is rarely required" bit now seems to refer to the wrong part.. The rarely wanted behaviour is the "all nodes get the same messages", NOT the shared subscriber functionality. 
As in it would be pretty weird to want all 4 nodes to each get the same messages to process - you'd almost always want to use the shared subscriber functionality to make sure each of the nodes gets different messages.

Reviewer 3: To avoid any kind of ambiguity, the connector could be removed and just say
"To avoid the message copies, there is a "shared subscriber" configuration....."

Reviewer 2: @eze210 
All Mule instances connect to the same JMS topic, which might lead to the repeated processing of messages when scaling the non-clustered Mule instance horizontally. To the JMS broker, the instances appear as separate subscribers, all of which get a copy of each message. This behavior is rarely required. To avoid this scenario,  a "shared subscriber" configuration is available to instruct the JMS broker to treat all instances as a combined subscriber and to give them each separate messages.

* JMS request-reply/request-response. +
All Mule instances are listening for messages in the same response queue. This implies that a Mule instance might obtain a response that isn't correlated to the request it sent and might result in incorrect responses or make a flow fail with timeout.

* Idempotent Redelivery Policy and Idempotent Message Validation. +

> Idempotency doesn’t work correctly with horizontal scaling if the same request is received by different Mule instances if the Object Store contents used by the Redelivery policy or the Idempotency Message Validator is localized. With the cluster sharing the Object Store values used by these idepotency features - duplicated messages aren’t possible because all nodes are sharing the list of already processed identifiers.

Reviewer: ```suggestion
Idempotency doesn’t work correctly with horizontal scaling if the same request is received by different Mule instances and the Object Store contents used by the Redelivery policy or the Idempotency Message Validator is localized. For a cluster sharing Object Store values used by these idempotency features, duplicate messages aren’t possible because all nodes are sharing the list of already-processed identifiers.
```

* Salesforce streaming API. +
If multiple instances of the same application are deployed, they will fail because the API only supports a single consumer. There is no failover support if the instance connected is stopped or crashes.

== Prerequisites

> * A cluster requires at least two Mule runtime engine instances, each one running on different machines (to avoid a single point of failure on the machine).

Reviewer: ```suggestion
* A cluster requires at least two Mule runtime engine instances, each one running on different machines to avoid a single point of failure on the machine.
```

All Mule instances get the same TCP requests and then process duplicate messages.

* JMS topics. +

> All Mule instances connect to the same JMS topic, which might lead to the repeated processing of messages when scaling the non-clustered Mule instance horizontally. To the JMS broker, the instances appear as separate subscribers, all of which get a copy of each message. This behavior is rarely required, but a "shared subscriber" configuration is available to instruct the JMS broker to treat all instances as a combined subscriber and to give them each separate messages.

Reviewer 2: ```suggestion
All Mule instances connect to the same JMS topic, which might lead to the repeated processing of messages when scaling the non-clustered Mule instance horizontally. To the JMS broker, the instances appear as separate subscribers, all of which get a copy of each message. This behavior is rarely required. To avoid this scenario, a "shared subscriber" configuration is available to instruct the JMS broker to treat all instances as a combined subscriber and to give them each separate messages.
```

== Cluster Design and Management

Anypoint Runtime Manager enables you to set up a customer-hosted cluster of Mule instances, and then deploy an application to run on the cluster. You can also monitor the status information for clusters and individual nodes. When clustered, you can easily manage several servers as one.

[NOTE]
For more detailed information about cluster management, see xref:runtime-manager::managing-servers.adoc[Managing Servers] in Runtime Manager.

> For Runtime Fabric an option exists at deployment time to provision the Mule runtime replicas in cluster mode.

Reviewer: ```suggestion
For Anypoint Runtime Fabric, an option exists at deployment time to provision the Mule runtime replicas in cluster mode.
```

A Mule Cluster consists of two or more Mule runtime engines, or nodes, grouped together and treated as a single unit. With the initial configuration, MuleSoft recommends that you scale a cluster to no more than eight Mule runtime engines.

With Anypoint Runtime Manager, you can deploy, monitor, or stop all the Mule runtime engines in a cluster as if they were a single Mule runtime engine.

> Note: CloudHub does not use this clustering configuration for provisioned workers, but it has other High Availability features to provide an equivalent experience, such as externalized state management via Anypoint Object Store (OSv2) or the persistent queues feature (which moves the VM queues external from the workers so they are shared and will survive re-provisioning).

Reviewer: Re "(which moves the VM queues external from the workers so they are shared and will survive re-provisioning)."

Question: moves queues that are already external to the workers or moves them outside of the workers so they can be shared ... ?

Author: VM Queues are locally stored on the worker (and thus not able to share messages beyond the worker) unless you tick the box "enable persistent queues" and then they will be external to the worker (worker 0 publishing could be then picked up by worker 1).

== Cluster Design and Management

Anypoint Runtime Manager enables you to set up a customer-hosted cluster of Mule instances, and then deploy an application to run on the cluster. You can also monitor the status information for clusters and individual nodes. When clustered, you can easily manage several servers as one.

[NOTE]
For more detailed information about cluster management, see xref:runtime-manager::managing-servers.adoc[Managing Servers] in Runtime Manager.

For Anypoint Runtime Fabric, an option exists at deployment time to provision the Mule runtime replicas in cluster mode.

A Mule Cluster consists of two or more Mule runtime engines, or nodes, grouped together and treated as a single unit. With the initial configuration, MuleSoft recommends that you scale a cluster to no more than eight Mule runtime engines.

With Anypoint Runtime Manager, you can deploy, monitor, or stop all the Mule runtime engines in a cluster as if they were a single Mule runtime engine.

> Note: CloudHub does not use this clustering configuration for provisioned workers, but it has other High Availability features to provide an equivalent experience, such as externalized state management via Anypoint Object Store (OSv2) or the persistent queues feature (which moves the VM queues external from the workers so they are shared and will survive re-provisioning).

Reviewer 2: ```suggestion
Note: CloudHub does not use this clustering configuration for provisioned workers, but it has other High Availability features to provide an equivalent experience, such as externalized state management via Anypoint Object Store (OSv2) or the persistent queues feature (which moves the VM queues outside the workers so they can be shared and survive re-provisioning).
```

=== Primary Node Difference

> In an active-active model, all nodes are able to perform processing. However, one of the nodes acts as the "primary node" which runs the schedulers and any event sources marked as "primary node only". This means that sources needing to run on just one node in a cluster can be configured to only be used by the primary polling node so that no other node reads messages from that source. 

Reviewer: ```suggestion
In an active-active model, all nodes are able to perform processing. However, one of the nodes acts as the primary node, which runs the schedulers and any event sources marked as "primary node only". This model enables you to configure sources to run on a single, primary polling node in a cluster and prevent other nodes in the cluster from reading messages from those sources. 
```

Reviewer: "primary node only": All event sources allow this setting? Any don't? use case?

Author: Scheduler doesn't have this tickbox/checkbox and perhaps other scheduler invoked stuff might not have it as an option because that's just how schedulers work. JMS's on new message is the only one I've come across that has it set to true by default.. e.g. HTTP listener has it unchecked/unticked.

== Clustering and Load Balancing

> When Mule clusters are used to serve TCP requests (where TCP includes SSL/TLS, UDP, Multicast, HTTP, and HTTPS), some load balancing is needed to distribute the requests among the clustered instances. While Runtime Fabric includes a load balancer capability as part of the underlying docker/K8s infrastructure, for customer-hosted/manually provisioned clusters you will need to supply a 3rd party load balancer or perform some sort of client-side load balancing/fail-over.

Reviewer: ```suggestion
When Mule clusters are used to serve TCP requests (where TCP includes SSL/TLS, UDP, Multicast, HTTP, and HTTPS), some load balancing is needed to distribute the requests among the clustered instances. Though Anypoint Runtime Fabric includes load-balancer capability as part of the underlying Docker Kubernetes (K8s) infrastructure, customer-hosted, manually-provisioned clusters require you to supply a third party load balancer or perform client-side load balancing and fail-over.
```
